Apply Force and Torque GUI plugin

Henrique-BO opened the pull request:
<!--
Please remove the appropriate section.
For example, if this is a new feature, remove all sections except for the "New feature" section

If this is your first time opening a PR, be sure to check the contribution guide:
https://gazebosim.org/docs/all/contributing
-->


# 🎉 New feature

Partially addresses #306 

## Summary
<!--Explain changes made, the expected behavior, and provide any other additional
context (e.g., screenshots, gifs) if appropriate.-->

Adds a new ApplyForceTorque GUI plugin, essentially implementing the [Apply Force and Torque](https://classic.gazebosim.org/tutorials?tut=apply_force_torque&cat=tools_utilities) interface that we have in Gazebo Classic.

This allows users to apply a specified force and torque to a chosen link in a model upon pressing a button. The implementation takes advantage of the existing ApplyLinkWrench system, sending the wrenches to the `/world/<world_name>/wrench` topic. The ApplyLinkWrench system is automatically loaded once this plugin is loaded.

However, a current limitation is that applying the force with an offset from the link origin is not supported, since it is not implemented in ApplyLinkWrench.

## Test it
<!--Explain how reviewers can test this new feature manually.-->

Open any test world.

```
gz sim src/gz-sim/examples/worlds/apply_link_wrench.sdf
```

Load the Apply Force Torque plugin from the plugin dropdown. 

On the interface, input the desired force and torque. The model is selected from the scene, and the specific link is specified through a dropdown on the interface. Buttons are available to apply either just a force, just a torque or both at the same time.

![ezgif com-video-to-gif](https://github.com/gazebosim/gz-sim/assets/49255495/a8134361-eab4-4c13-bbb5-7c8aa0e72928)

## Checklist
- [x] Signed all commits for DCO
- [ ] Added tests
- [ ] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] Consider updating Python bindings (if the library has them)
- [x] `codecheck` passed (See [contributing](https://gazebosim.org/docs/all/contributing#contributing-code))
- [x] All tests passed (See [test coverage](https://gazebosim.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Agazebosim+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` messages.

azeey commented on src/gui/plugins/apply_force_torque/ApplyForceTorque.qml:
```diff
@@ -0,0 +1,300 @@
+      forceX.value, forceY.value, forceZ.value)
+  }
+
+  Label {
```
I think it's best to leave out the offset portion until we actually support it.

Henrique-BO replied:
Should we remove it entirely from the interface for now or just block the user from inputing values (like with a `enabled: false`)?

azeey commented on src/gui/plugins/apply_force_torque/ApplyForceTorque.qml:
```diff
@@ -0,0 +1,300 @@
+    Layout.columnSpan: 2
+    Layout.fillWidth: true
+    id: forceX
+    maximumValue: 1000000
```
Can we make these constants, so we can change them easily if needed?

azeey commented on src/gui/plugins/apply_force_torque/ApplyForceTorque.qml:
```diff
@@ -0,0 +1,300 @@
+      ApplyForceTorque.linkIndex = currentIndex
+    }
+    // ToolTip.visible: hovered
+    // ToolTip.text: qsTr("Link to which the wrench should be applied")
```
Remove?

azeey commented on src/gui/plugins/apply_force_torque/ApplyForceTorque.qml:
```diff
@@ -0,0 +1,300 @@
+    Layout.columnSpan: 6
+    id: linkCombo
+    Layout.fillWidth: true
+    // enabled: ApplyForceTorque.configured
```
Remove?

azeey commented on src/gui/plugins/apply_force_torque/ApplyForceTorque.qml:
```diff
@@ -0,0 +1,300 @@
+import QtQuick.Dialogs 1.0
+import QtQuick.Controls.Material 2.1
+import QtQuick.Layouts 1.3
+import "qrc:/qml"
```
Is this needed?

azeey replied:
Can you address this?

Henrique-BO replied:
The current version has only the necessary imports

azeey commented on src/gui/plugins/apply_force_torque/ApplyForceTorque.cc:
```diff
@@ -0,0 +1,336 @@
+      // Put all of the model's links into the list
+      auto links = selectedModel.Links(_ecm);
+      unsigned int i{0};
+      while (i < selectedModel.LinkCount(_ecm))
```
nitpick:
```suggestion
      while (i < links.size())
```
since we're actually indexing into `links`.

azeey commented on src/gui/plugins/apply_force_torque/ApplyForceTorque.cc:
```diff
@@ -0,0 +1,336 @@
+      }
+
+      // Create publisher if not yet created
+      if (!this->dataPtr->pub.Valid())
+      {
+        std::string worldName{""};
+        _ecm.Each<components::World, components::Name>(
+            [&](const Entity &_entity,
+              const components::World * /* _world */ ,
+              const components::Name *_name)->bool
+            {
+              World world(_entity);
+              for (auto &model : world.Models(_ecm))
+              {
+                if (model == entity)
+                {
+                  worldName = _name->Data();
+                  return true;
+                }
+              }
+              gzerr << "World not found" << std::endl;
+              return false;
+            });
+        auto topic = transport::TopicUtils::AsValidTopic(
+          "/world/" + worldName + "/wrench");
+        this->dataPtr->pub =
+            this->dataPtr->node.Advertise<msgs::EntityWrench>(topic);
+        gzdbg << "Created publisher to " << topic << std::endl;
```
Creation of the publisher should be done in `LoadConfig` and you can get the worldName using `gui::worldNames` similar to https://github.com/gazebosim/gz-sim/blob/064f69ed7c441b90abbf0f84eeb0bfaeec628da9/src/gui/plugins/spawn/Spawn.cc#L186-L188

azeey commented on src/gui/plugins/apply_force_torque/ApplyForceTorque.cc:
```diff
@@ -0,0 +1,336 @@
+  {
+    gz::sim::gui::events::EntitiesSelected *_e =
+        static_cast<gz::sim::gui::events::EntitiesSelected*>(_event);
+    this->dataPtr->selectedEntities = _e->Data();
```
If we only need one selected entity, can we just store that instead of the vector? If you need to know whether the selected entity has been set or not, you can use `std::optional`.

azeey commented on src/gui/plugins/apply_force_torque/ApplyForceTorque.cc:
```diff
@@ -0,0 +1,336 @@
+    }
+  }
+
+  this->ModelNameChanged();
```
Do you need `emit` for these?

azeey commented on src/gui/plugins/apply_force_torque/ApplyForceTorque.qml:
```diff
@@ -0,0 +1,300 @@
+    text: qsTr("X (N)")
+  }
+
+  GzSpinBox {
```
Can we have a message here that the force would be applied at the link origin, and not the COM?

azeey reviewed (commented):
Looks great!! I haven't tested it yet, but wanted to give it a first pass.

azeey commented on src/gui/plugins/apply_force_torque/ApplyForceTorque.cc:
```diff
@@ -0,0 +1,336 @@
+    public: transport::Node::Publisher pub;
+
+    /// \brief To synchronize member access
+    public: std::mutex mutex;
```
Since everything here is happening from a single Qt thread, I don't think you need this mutex.

azeey replied:
alas, I was wrong about this. The `EntitiesSelected` and `DeselectAllEntities` can come from the main Qt thread (via the `EntityTree` plugin) as well as the rendering thread (via the `SelectEntities` plugin), so a mutex is necessary. Sorry about that :(

arjo129 commented on src/gui/plugins/apply_force_torque/ApplyForceTorque.qml:
```diff
@@ -0,0 +1,300 @@
+      ApplyForceTorque.ApplyAll()
+    }
+  }
+}
```
Knit: New line at end of file.

arjo129 commented on src/gui/plugins/apply_force_torque/ApplyForceTorque.hh:
```diff
@@ -0,0 +1,132 @@
+  /// \brief Publish wrench to "/world/apply_link_wrench/wrench" topic.
+  ///
+  /// ## Configuration
+  /// This plugin doesn't accept any custom configuration.
```
Knit: Might be worth mentioning which plugin provides the `/world/apply_link_wrench/wrench` topic.

arjo129 reviewed (commented):
Minor documentation related feedback. Which plugin is needed in order to apply the force?

ahcorde commented on src/gui/plugins/apply_force_torque/ApplyForceTorque.cc:
```diff
@@ -0,0 +1,336 @@
+              gzerr << "World not found" << std::endl;
+              return false;
+            });
+        auto topic = transport::TopicUtils::AsValidTopic(
```
check if topic is valid. it should return an empty string if the function fails

ahcorde commented on src/gui/plugins/apply_force_torque/ApplyForceTorque.cc:
```diff
@@ -0,0 +1,336 @@
+#include <gz/msgs/entity_wrench.pb.h>
+#include <gz/plugin/Register.hh>
+#include <gz/transport/Node.hh>
+
```
include `gz/mah/Vector3.hh`,

quarkytale reviewed (commented):
Tried it out, works as expected

Henrique-BO commented:
Thanks for the feedback. I addressed the requested changes in 6fbbfc0.

quarkytale commented:
@osrf-jenkins run tests please

azeey commented on src/gui/plugins/apply_force_torque/ApplyForceTorque.cc:
```diff
@@ -0,0 +1,297 @@
+
+  gz::gui::App()->findChild<gz::gui::MainWindow *>
+    ()->installEventFilter(this);
+  gz::gui::App()->findChild<gz::gui::MainWindow *>
+    ()->QuickWindow()->installEventFilter(this);
```
I'm not sure if we need to install the event filter on QuickWindow.

azeey commented on src/gui/plugins/apply_force_torque/ApplyForceTorque.qml:
```diff
@@ -0,0 +1,243 @@
+    Layout.columnSpan: 8
+    id: forceText
+    color: "dimgrey"
+    text: qsTr("Force (applied to link origin):")
```
I checked how Gazebo-classic works and there, the application point is the center of mass. Would it be possible to do the same here?

ahcorde commented on src/gui/plugins/apply_force_torque/ApplyForceTorque.cc:
```diff
@@ -0,0 +1,297 @@
+
+#include <string>
+
+#include <gz/gui/Application.hh>
```
alphabetize headers

ahcorde commented on src/gui/plugins/apply_force_torque/ApplyForceTorque.cc:
```diff
@@ -0,0 +1,297 @@
+  {
+    this->dataPtr->changedEntity = false;
+
+    this->dataPtr->modelName = "";
```
```suggestion
    this->dataPtr->modelName{""};
```

Henrique-BO replied:
I don't think this works, since this is an assignment, not an initialization. I get a compilation error if I try this.

ahcorde commented on src/gui/plugins/apply_force_torque/ApplyForceTorque.hh:
```diff
@@ -0,0 +1,127 @@
+#define GZ_GUI_APPLYFORCETORQUE_HH_
+
+#include <memory>
+
```
include this headers `QString`, `QStringList`. `Qobject`, `QEvent`, 

`#include <ignition/gazebo/EntityComponentManager.hh>`

Henrique-BO commented:
Changes addressed in ca1c904.

azeey commented on src/gui/plugins/apply_force_torque/ApplyForceTorque.cc:
```diff
@@ -0,0 +1,321 @@
+  math::Vector3d forceToApply =
+    _applyForce ? this->force : math::Vector3d::Zero;
+  math::Vector3d torqueToApply =
+    _applyTorque ? this->torque : math::Vector3d::Zero +
```
Parenthesize the ternary expression. Otherwise, I'm not sure if the cross product term is being applied in both conditions.

azeey reviewed (commented):
I checked gazebo-classic once more and it looks like the force/torque is applied in the body fixed frame. I don't want to block this PR on that. Maybe we can add that later, but I think we should indicate that the force/torque is applied in the world frame in the GUI.

Henrique-BO commented:
You're right, I hadn't noticed that. I changed the application to the body fixed frame and added an indication to the GUI, so it's more consistent with Gazebo-Classic.

azeey commented on src/gui/plugins/apply_force_torque/ApplyForceTorque.cc:
```diff
@@ -0,0 +1,318 @@
+}
+
+// Register this plugin
+GZ_ADD_PLUGIN(ApplyForceTorque, gz::gui::Plugin);
```
The `;` is causing a warning.
```suggestion
GZ_ADD_PLUGIN(ApplyForceTorque, gz::gui::Plugin)
```

Henrique-BO replied:
Done in 3911d73
